src/typechecker/define-instance.lisp

```diff
@@ -28,17 +29,52 @@
 
 (in-package #:coalton-impl/typechecker/define-instance)
 
+(defun expand-constraint (base-constraint env)
+  "This was implemented as a hack to make `derive' work on recursive
```

Comment:
Describe what the function does in the docstring, and add commentary about hacks and purpose in comments.